packages/ember-application/lib/system/application-instance.js

```diff
@@ -212,6 +298,255 @@ let ApplicationInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
+      returns a promise that resolves with the app instance when the transition
+      is complete, or rejects if the transion was aborted due to an error.
+
+      @public
+      @param url {String} the destination URL
+      @return {Promise}
+    */
+    visit(url) {
+      this.setupRouter();
+
+      let router = get(this, 'router');
+
+      let handleResolve = () => {
+        // Resolve only after rendering is complete
+        return new Ember.RSVP.Promise((resolve) => {
+          // TODO: why is this necessary? Shouldn't 'actions' queue be enough?
+          // Also, aren't proimses supposed to be async anyway?
+          run.next(null, resolve, this);
+        });
+      };
+
+      let handleReject = (error) => {
+        if (error.error) {
+          throw error.error;
+        } else if (error.name === 'TransitionAborted' && router.router.activeTransition) {
+          return router.router.activeTransition.then(handleResolve, handleReject);
+        } else if (error.name === 'TransitionAborted') {
+          throw new Error(error.message);
+        } else {
+          throw error;
+        }
+      };
+
+      // Keeps the location adapter's internal URL in-sync
+      get(router, 'location').setURL(url);
+
+      return router.handleURL(url).then(handleResolve, handleReject);
+    }
+  });
+
+  /**
+    A list of boot-time configuration options for customizing the behavior of
+    an `Ember.ApplicationInstance`.
+
+    This is an interface class that exists purely to document the available
+    options; you do not need to construct it manually. Simply pass a regular
+    JavaScript object containing the desired options into methods that require
+    one of these options object:
+
+    ```javascript
+    MyApp.visit("/", { location: "none", rootElement: "#container" });
+    ```
+
+    Not all combinations of the supported options are valid. See the documentation
+    on `Ember.Application#visit` for the supported configurations.
+
+    Internal, experimental or otherwise unstable flags are marked as private.
+
+    @class BootOptions
+    @namespace @Ember.ApplicationInstance
+    @public
+  */
+  BootOptions = function BootOptions(options) {
+    options = options || {};
+
+    /**
+      Provide a specific instance of jQuery. This is useful in conjunction with
+      the `document` option, as it allows you to use a copy of `jQuery` that is
+      appropiately bound to the foriegn `document` (e.g. a jsdom).
+
+      This is highly experimental and support very incomplete at the moment.
+
+      @property jQuery
+      @type Object
+      @default auto-detected
+      @private
+    */
+    this.jQuery = jQuery;
+
+    /**
+      Interactive mode: whether we need to set up event delegation and invoke
+      lifecycle callbacks on Components.
+
+      @property interactive
+      @type boolean
+      @default auto-detected
+      @private
+    */
+    this.interactive = environment.hasDOM;
```

Comment:
This is `destinedForDOM` AND `please set up event delegation` (event dispatcher).. it wanted to consolidate the two into a single thing since it doesn't seem to make sense to have one but not the other, thoughts?